model.py

```diff
@@ -1,11 +1,25 @@
+
+import itertools
 import torch as th
-from typing import Dict, Tuple, Union
+from typing import Dict, Tuple, Union, Iterable
 nn = th.nn
 F = nn.functional
+import einops
+
+
+def pairwise(seq: Iterable):
+    """s -> (s0,s1), (s1,s2), (s2, s3), ..."""
+    a, b = itertools.tee(seq)
+    next(b, None)
+    return zip(a, b)
+
+
+def epairwise(seq: Iterable):
+    return enumerate(pairwise(seq))
```

Comment:
These utility functions may be moved into a separate file later.

Reply:
ok